Document MSTestAnalysisMode property (#44632)
dotnet/docs: docs/core/testing/mstest-analyzers/overview.md

Reviewer: Should we also add the property to https://github.com/dotnet/docs/blob/5802e8b74127b0d9d2a116d5be43674d78179c30/docs/core/project-sdk/msbuild-props.md?plain=1#L1442?

The rules are organized into categories such as performance usage...

Starting with MSTest.TestFramework 3.7, the MSTest.Analyzers NuGet package is a dependency of the framework. For earlier versions, you need to use `MSTest` metapackage or add a package reference for `MSTest.Analyzers` explicitly.

## MSTestAnalysisMode

Starting with MSTest 3.8, an MSBuild property named `MSTestAnalysisMode` is available to determine which analyzers are enabled at which severity.

The available values are:

### `None`

This sets all analyzers to `none` severity, making all of them disabled. You can then enable individual analyzers using `.editorconfig` or `.globalconfig` files.

### `Default`

This follows the default documented behavior for each rule.

- Rules that are enabled by default will use their default severity.
- Rules that are disabled by default will use `none` severity.

> ### `Recommended`

Reviewer 2: It might be a bit too boring to maintain the doc but it would probably be awesome if we could list here the rules enabled by this mode.

Reviewer 2: Maybe we could generate a markdown per version and we would include it here.
@IEvangelist is this something that is common?

Reviewer 2: @gewarren what's your opinion here?

Reviewer 2: Given we have NuGet not part of SDK or VS, it'd be more complex for us. But I think just pointing out to the file in our NUGet package is a good idea.

Author: @gewarren @Evangelink I added some info on how to locate the globalconfig inside the NuGet package. Let me know if it looks good

> This mode is more aggressive than `Default`. Rules that are enabled by default with Info (`suggestion`) severity are escalated to be warnings. Moreover, certain rules may decide that they are escalated to be errors in both `Recommended` and `All` modes. For example, [MSTEST0003: Test methods should have valid layout](mstest0003.md) is escalated to error in `Recommended` and `All` modes.

Reviewer 2: The first sentence could feel a little scary to users. I would instead say that in `Default` we are ensuring to not have rules that would cause warning or error and that don't result in runtime error. Recommended is what we think most dev should use but this means that a minor update of the package could cause new warnings.

Author: > Recommended is what we think most dev should use but this means that a minor update of the package could cause new warnings.

But a minor update can cause new warnings for "Default" as well?

Reviewer 2: Only for cases where there is an execution error. Otherwise we mark the new rules as info

### `All`

This mode is more aggressive than `All`. All rules are enabled as warnings. As mentioned for `Recommended` mode, certain rules may decide that they are escalated to be errors in both `Recommended` and `All` modes. For example, [MSTEST0003: Test methods should have valid layout](./mstest0003.md) is escalated to error in `Recommended` and `All` modes.

> [!NOTE]
> The following rules are completely opt-in and are not enabled in `Default`, `Recommended`, or `All` modes:
>
> - [MSTEST0015: Test method should not be ignored](mstest0015.md)
> - [MSTEST0019: Prefer TestInitialize methods over constructors](mstest0019.md)
> - [MSTEST0020: Prefer constructors over TestInitialize methods](mstest0020.md)
> - [MSTEST0021: Prefer Dispose over TestCleanup methods](mstest0021.md)

> > - [MSTEST0022: Prefer TestCleanup over Dispose methods](mstest0022.md)

Author: Not sure where to best document this. Options:

1. In this page only as I did in this PR
2. In doc page of the specific analyzer, with only a note here mentioning "some rules" may have this behavior
3. Both as I did here plus in the specific analyzer page.

Reviewer 2: Probably 3 but the specific analyzer page update can be done on a follow-up PR IMO.

Reviewer 2: Do you want to address this now or later?

Author: Done

## MSTestAnalysisMode

Starting with MSTest 3.8, an MSBuild property named `MSTestAnalysisMode` is available to determine which analyzers are enabled at which severity.

> [!TIP]
> To see which rules are enabled at which severity for each mode, you can navigate to the package of the version of interest in NuGet cache, locate the `globalconfigs` directory, and open the `.globalconfig` file corresponding to the analysis mode.
> For more information on finding the NuGet cache directory, see [Managing the global packages, cache, and temp folders](/nuget/consume-packages/managing-the-global-packages-and-cache-folders). From that directory, locate `mstest.analyzers` directory, then the version (3.8 and higher), then `globalconfigs`.
> Alternatively, you can download the NuGet package of the version of interest from `nuget.org` and view it in NuGet Package Explorer (Windows app), or view directly in the [web app version of NuGet Package Explorer](https://nuget.info/packages/MSTest.Analyzers/).

> The available values for this property:

Reviewer: Do you want to manually create an index to the H3s here?

Reviewer 2: What do you mean?

Reviewer: ```suggestion
The available values for this property:

- [`None`](#none)
- [`Default`](#default)
- [`Recommended`](#recommended)
- [`All`](#all)
```

Reviewer 2: Aaaah! Right sure, it can be useful.

### `None`

> This sets all analyzers to `none` severity, making all of them disabled. You can then enable individual analyzers using `.editorconfig` or `.globalconfig` files.

Reviewer: ```suggestion
This value sets all analyzers to `none` severity, disabling all of them. You can then enable individual analyzers using `.editorconfig` or `.globalconfig` files.
```

### `Default`

> This follows the default documented behavior for each rule.

Reviewer: ```suggestion
This setting follows the default documented behavior for each rule.
```

- Rules that are enabled by default will use their default severity.
- Rules that are disabled by default will use `none` severity.

> [!NOTE]

> > Rules that are enabled by default as warnings are violations which are expected to cause issues at runtime.

Reviewer: ```suggestion
> Rules that are enabled by default as warnings are violations that are expected to cause issues at run time.
```

> > Rules that are enabled by default as warnings are violations which are expected to cause issues at run-time.

Reviewer 2: ```suggestion
> Rules that are enabled by default as warnings are violations which are expected to cause issues at runtime.
```

### `Recommended`

> This is the mode we expect most developers to use. Rules that are enabled by default with Info (`suggestion`) severity are escalated to be warnings. Moreover, certain rules may decide that they are escalated to be errors in both `Recommended` and `All` modes. For example, [MSTEST0003: Test methods should have valid layout](mstest0003.md) is escalated to error in `Recommended` and `All` modes.

Reviewer: ```suggestion
This is the mode we expect most developers to use. Rules that are enabled by default with Info (`suggestion`) severity are escalated to warnings. Moreover, certain rules might be escalated to errors in both `Recommended` and `All` modes. For example, [MSTEST0003: Test methods should have valid layout](mstest0003.md) is escalated to error in `Recommended` and `All` modes.
```

### `All`

> This mode is more aggressive than `Recommended`. All rules are enabled as warnings. As mentioned for `Recommended` mode, certain rules may decide that they are escalated to be errors in both `Recommended` and `All` modes. For example, [MSTEST0003: Test methods should have valid layout](./mstest0003.md) is escalated to error in `Recommended` and `All` modes.

Reviewer: ```suggestion
This mode is more aggressive than `Recommended`. All rules are enabled as warnings. As mentioned for `Recommended` mode, certain rules might be escalated to errors in both `Recommended` and `All` modes. For example, [MSTEST0003: Test methods should have valid layout](./mstest0003.md) is escalated to error in `Recommended` and `All` modes.
```
